### DISCUSSION TITLE
How would you type a Record Object in MongoDB ?

### DISCUSSION THREAD

**Discussion by maxime4000** [prisma/prisma#17963] [Q&A] [answered]
So I need to logs data coming from an other existing systems. 
The data is sent like a [Record<Key,Type>](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type) 
```json
{
  "device": {
    "1": { ... },
    "2": { ... },
    "4": { ... },
    "20": { ... }
  }
}
```

How would you do your schemas ?
```prisma
type DeviceData { ... }
type Device {
  // ???
}

model Data {
  id String @id @default(auto()) @map("_id") @db.ObjectId
  device Device
  // ...
}
```

**Comment by maxime4000** [comment, marked as answer]
I can always restructure my data before inserting it, but it's valid in MongoDB so I'm asking if this could be possible ?

Answer : Too complex for no real reason. Alternative exist, so no reason to do this.

**Reply by nurul3101** [reply]
Are the 20 entries fixed in the payload, or is the number dynamic?

**Reply by maxime4000** [reply]
Dynamic, those would be the equivalent of Ids. Like said, I can always map the data to an other format before inserting them, but I'm curious if this could still be possible.

**Reply by nurul3101** [reply]
Dynamic types wouldn't be possible, you would need to use JSON type to add dynamic key value pairs.

**Reply by maxime4000** [reply]
Sad, but I think it make sense to not support this. Thanks for your answers. :)

**Reply by nurul3101** [reply]
I would also recommend to try out [prisma-json-types-generator](https://github.com/arthurfiorette/prisma-json-types-generator) to add types for the JSON field.

**Comment by arthurfiorette** [comment]
This kind of complex modelling isn't supported by the prisma schema.

However, after some development I was able to create [prisma-json-types-generator](https://github.com/arthurfiorette/prisma-json-types-generator) focused on typing [Json for postgres](https://github.com/prisma/prisma/issues/3219) but should work for mongodb too.


This example in its readme should clarify the idea:

```prisma
// schema.prisma

generator client {
  provider = "prisma-client-js"
}

/// Always after the prisma-client-js generator
generator json {
  provider = "prisma-json-types-generator"
  // namespace = "PrismaJson"
  // clientOutput = "<finds it automatically>"
  // (./ -> relative to schema, or an importable path to require() it)
}

model Example {
  /// [MyType]
  normal Json

  /// [MyType]
  optional Json?

  /// [MyType]
  array Json[]
}
```

```ts
// index.ts

import type { Example } from '@prisma/client';

declare global {
  namespace PrismaJson {
    // you can use classes, interfaces, types, etc.
    type MyType = boolean;
  }
}

function myFunction(example: Example) {
  // example.normal   is now a boolean
  // example.optional is now a boolean | null
  // example.array    is now a boolean[]
}
```

**Comment by lmssiehdev** [comment]
@maxime4000 did you find a solution for this? Prisma doesn't seem to support objects yet.

**Reply by nurul3101** [reply]
As Arthur mentioned, you can have a look at `prisma-json-types-generator` and in case you have any further queries, feel free to open a new GitHub Discussion.

**Comment by arthurfiorette** [comment]
Give a look at `prisma-json-types-generator`
